app/code/Magento/Catalog/Model/ResourceModel/Product/Indexer/Eav/AbstractEav.php

```diff
@@ -148,6 +148,21 @@ public function reindexAttribute($attributeId, $isIndexable = true)
     abstract protected function _prepareIndex($entityIds = null, $attributeId = null);
 
     /**
+     * Prepare product visibility attribute value index
+     *
+     * @param array $entityIds      the entity ids limitation
+     * @return $this
+     */
+    protected function _prepareVisibilityIndex($entityIds = null)
+    {
+        $attribute = $this->_eavConfig->getAttribute(\Magento\Catalog\Model\Product::ENTITY, 'visibility');
+        $this->_prepareIndex($entityIds, $attribute->getAttributeId());
```

Comment:
why do we have additional call for _prepareIndex? if we had one before in chaining call.

This code looks awkward for me, as we have:

$this->_prepareIndex()->_prepareRelationIndex()->_prepareIndex($entityIds, $attribute->getAttributeId())->_removeNotVisibleEntityFromIndex();

The fact that we should to call the same method twice says about code smell